oauthproxy.go

```diff
@@ -699,6 +726,16 @@ func (p *OAuthProxy) OAuthStart(rw http.ResponseWriter, req *http.Request) {
 		return
 	}
 
+	// A session is created with `Authenticated: false` to store the code verifier
+	// for token redemption
+	session := &sessionsapi.SessionState{CodeVerifier: codeVerifier}
+	err = p.SaveSession(rw, req, session)
+	if err != nil {
+		logger.Errorf("Error saving session state for %s: %v", req.RemoteAddr, err)
+		p.ErrorPage(rw, req, http.StatusInternalServerError, err.Error())
+		return
+	}
```

Comment:
Could this not be encoded as part of the state parameter that we send through the IDP? That's what we are doing with the OIDC nonce for example

Reply:
Interesting. I'm not familiar with how you are handling the OIDC nonce. Are you encrypting this in the state paramter?

I'm not sure that this is a very "safe" way to handle this. The `code_verifier` should never be sent outside of oauth2-proxy until authorization code redemption. It is considered a secret until that point.

Reply:
If I recall, if you peek at the `csrf` package, we have the OAuth2 state and OIDC nonce inside our CSRF cookie for traversing the initial sign in flow -- the contents are encrypted with the OAuth2-Proxy secret.

Reply:
`cookie` package under `pkg/cookies/csrf` -- take a peek at the functions there around the `CSRF` interface. The `csrf` struct that has state/nonce is AES-CFB encrypted (our choice for cookies to avoid AES-GCM collision issues) and SHA256 signed.

Reply:
Do you both believe it is better to send the code verifier back and forth in the request? 
My concerns are possibly leaking the code verifier. The whole point of PKCE is so that only the client knows the actual code and isn't leaked to the browser or OS in anyway. In addition it increases the size of the payload contained in the csrf cookie.

I'm actually a bit surprised that the CSRF cookie contains encrypted nonce data

Reply:
I'll defer to you the best mechanism to store details you need across request state before a session exists (if you need that) 😄 

I only mention the CSRF cookie since we have ready structure & support for storing state/nonce with encryption and signing if needed.

I think I just used the `csrf` cookie because it already existed for state checks & the nonce check for OIDC followed a nearly identical flow path. Here's the PR for context on what it did (and how it fixed up previous csrf cookie issues: https://github.com/oauth2-proxy/oauth2-proxy/pull/967

> In addition it increases the size of the payload contained in the csrf cookie.

No need to worry about that! Today it just has the 32byte state/nonce fields + whatever the signature + timestamp length is. Still plenty of headroom before you'd hit the 4096 character cookie value max 😄

Reply:
> The whole point of PKCE is so that only the client knows the actual code and isn't leaked to the browser or OS in anyway

In case that's a concern as you are choosing a design path: All our cookies are completely encrypted with the `secret` config option (session & csrf) in addition to being SHA256 signed (with that secret). Why encrypt + signature? We use AES-CFB in cookie based sessions to remove the risk of IV collisions AES-GCM has (low risk, would need 2billion sessions with the same secret - but better safe than sorry 😄 ). But since CFB isn't authenticated like GCM by default, we sign the cookies ourselves.